### PR TITLE
Renesas: Smartbond: Add MIPI DBI Driver Support

### PR DESCRIPTION
**Pull request by ioannis-karachalios**

### CODE REVIEW

File: boards/arm/da1469x_dk_pro/da1469x_dk_pro-pinctrl.dtsi
Comment (danieldegrasse) [inline boards/arm/da1469x_dk_pro/da1469x_dk_pro-pinctrl.dtsi:1]: Please do not include the pin control changes in the commit adding the driver- group them with changes for the board

File: drivers/mipi_dbi/mipi_dbi_smartbond.c [hunk 1]
```diff
@@ -0,0 +1,610 @@
...
+	/* Reset GPIO */
+	const struct gpio_dt_spec reset;
+	/* Host controller's timing settings */
+	lcdc_smartbond_timing_cfg timing_cfg;
+	/* Background default color configuration */
+	lcdc_smartbond_bgcolor_cfg bgcolor_cfg;
+};
+
+/* Mark the device is is progress and so it's not allowed to enter the sleep state. */
+static void mipi_dbi_pm_get(const struct device *dev)
```
Comment (danieldegrasse) [inline drivers/mipi_dbi/mipi_dbi_smartbond.c:81]: Not sure if this will work for your use case, but `pm_policy_state_lock_get` might also be helpful here

File: drivers/mipi_dbi/mipi_dbi_smartbond.c [hunk 2]
```diff
@@ -0,0 +1,610 @@
...
+#include <zephyr/drivers/display.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/drivers/spi.h>
+#include <da1469x_lcdc.h>
+#include <da1469x_pd.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(smartbond_mipi_dbi, CONFIG_MIPI_DBI_LOG_LEVEL);
+
+#define SMARTBOND_IRQN		DT_INST_IRQN(0)
```
Comment (danieldegrasse) [inline drivers/mipi_dbi/mipi_dbi_smartbond.c:29]: Generally speaking, I'd prefer that this driver be multi-instance, unless there is a good reason not to make it support multiple device instances. These properties should likely be part of the `dev->config` structure

File: drivers/mipi_dbi/mipi_dbi_smartbond.c [hunk 3]
```diff
@@ -0,0 +1,610 @@
...
+	}
+}
+
+#if MIPI_DBI_SMARTBOND_IS_RESET_REQUIRED
+static int mipi_dbi_smnartbond_reset(const struct device *dev, uint32_t delay)
+{
+	const struct mipi_dbi_smartbond_config *config = dev->config;
+	int ret;
+
+	if (config->reset.port == NULL) {
```
Comment (danieldegrasse) [inline drivers/mipi_dbi/mipi_dbi_smartbond.c:160]: Could you use `gpio_is_ready_dt` here?

File: drivers/mipi_dbi/mipi_dbi_smartbond.c [hunk 4]
```diff
@@ -0,0 +1,610 @@
...
+		LOG_ERR("Cannot drive reset signal");
+		return ret;
+	}
+	k_msleep(delay);
+
+	return gpio_pin_set_dt(&config->reset, 0);
+}
+#endif
+
+/* Dispplay pixel to output color format translation */
```
Comment (danieldegrasse) [inline drivers/mipi_dbi/mipi_dbi_smartbond.c:176]: Nit: Dispplay -> Display

File: drivers/mipi_dbi/mipi_dbi_smartbond.c [hunk 5]
```diff
@@ -0,0 +1,610 @@
...
+{
+	mipi_dbi_cfg->cpha = dbi_config->config.operation & SPI_MODE_CPHA;
+	mipi_dbi_cfg->cpol = dbi_config->config.operation & SPI_MODE_CPOL;
+	mipi_dbi_cfg->cs_active_high = dbi_config->config.operation & SPI_CS_ACTIVE_HIGH;
+	mipi_dbi_cfg->line_mode = lcdc_smartbond_line_mode_translation(dbi_config->mode);
+	mipi_dbi_cfg->color_mode = lcdc_smartbond_pixel_to_ocm(pixfmt);
+}
+
+#if MIPI_DBI_SMARTBOND_IS_READ_SUPPORTED
+static int mipi_dbi_smartbond_command_read(const struct device *dev,
```
Comment (danieldegrasse) [inline drivers/mipi_dbi/mipi_dbi_smartbond.c:230]: Just a question here- do you want to support reads? Your driver, and totally up to you. I don't envision every controller supporting reads though, because this is often not a use case required by users- so I just want to make it clear that if you don't see a need for this support, you are welcome to simply not implement it, since the hardware doesn't enable support.

File: drivers/mipi_dbi/mipi_dbi_smartbond.c [hunk 6]
```diff
@@ -0,0 +1,610 @@
...
+		goto _mipi_dbi_write_exit;
+	}
+
+	ret = da1469x_lcdc_timings_configure(desc->width, desc->height,
+			(lcdc_smartbond_timing_cfg *)&config->timing_cfg);
+	if (ret < 0) {
+		goto _mipi_dbi_write_exit;
+	}
+
+	LCDC_SMARTBOND_LAYER_CONFIG(layer, framebuf, 0, 0, desc->width, desc->height,
```
Comment (danieldegrasse) [inline drivers/mipi_dbi/mipi_dbi_smartbond.c:393]: You should validate that the `buf_size` field matches the `height` and `width` of the descriptor here, because the MIPI DBI API specifies that `buf_size` is the amount of data the controller will send , not `height` and `width`. I chose this distinction because most MIPI DBI controllers likely only care about how many bytes to stream out, which can be determined using `buf_size`.

*Note- the `buf_size` restriction isn't currently documented well. #68378 should clarify this*

File: drivers/mipi_dbi/mipi_dbi_smartbond.c [hunk 7]
```diff
@@ -0,0 +1,610 @@
...
+	(((_var) << LCDC_LCDC_LAYER0_OFFSETX_REG_ ## _field ## _Pos) & \
+	LCDC_LCDC_LAYER0_OFFSETX_REG_ ## _field ## _Msk)
+
+struct mipi_dbi_smartbond_data {
+	/* Provide mutual exclusion when a display operation is requested. */
+	struct k_sem device_sem;
+	/* Provide synchronization between task return and ISR firing */
+	struct k_sem sync_sem;
+	/* Flag indicating whether or not an underflow took place */
+	bool underflow_flag;
```
Comment (danieldegrasse) [inline drivers/mipi_dbi/mipi_dbi_smartbond.c:64]: Just a question- does this need to be `volatile`? I suppose the compiler will likely always read this from memory, but might be worth being explicit since this is touched in an ISR.

File: drivers/mipi_dbi/mipi_dbi_smartbond.c [hunk 8]
```diff
@@ -0,0 +1,610 @@
...
+	.reset = mipi_dbi_smnartbond_reset,
+#endif
+	.command_write = mipi_dbi_smartbond_command_write,
+	.write_display = mipi_dbi_smartbond_write_display,
+#if MIPI_DBI_SMARTBOND_IS_READ_SUPPORTED
+	.command_read = mipi_dbi_smartbond_command_read,
+#endif
+};
+
+#define SMARTBOND_MIPI_DBI_INIT(inst)	\
```
Comment (danieldegrasse) [inline drivers/mipi_dbi/mipi_dbi_smartbond.c:588]: I mentioned this above- but we should make this driver support multiple instances. The macros here are already instance based, so if there is a compelling reason to keep the driver single instance we need to use single instance macros here as well.

File: dts/bindings/mipi-dbi/renesas,smartbond-mipi-dbi.yaml [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+  te-inversion:
+    type: boolean
+    description: |
+      Boolean to apply an inversion on the TE signal that triggers the MIPI DBI controller.
+
+  dma-prefetch:
+    type: int
+    enum:
+      - 0 # Prefetch mechanism is disabled
```
Comment (danieldegrasse) [inline dts/bindings/mipi-dbi/renesas,smartbond-mipi-dbi.yaml:35]: You can use an enum type here of `string` and then use `DT_INST_ENUM_IDX` in the code to get the value you need, this will be easier for the user to follow

File: modules/lvgl/Kconfig
Comment (danieldegrasse) [inline modules/lvgl/Kconfig:1]: I think we want this change in a separate PR, since it touches generic code.

File: samples/modules/lvgl/demos/boards/da1469x_dk_pro.conf [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+#
+#
+CONFIG_HEAP_MEM_POOL_SIZE=4096
+CONFIG_LOG_MODE_MINIMAL=y
+CONFIG_MIPI_DBI_LOG_LEVEL_DBG=y
+CONFIG_DISPLAY_LOG_LEVEL_DBG=n
+CONFIG_INPUT_FT5336_INTERRUPT=y
+CONFIG_LV_Z_POINTER_INPUT_MSGQ_COUNT=70
+# Frame buffer must always be 4-byte aligned. This is imposed by LCDC.
+CONFIG_LV_Z_VDB_ALIGN=4
```
Comment (danieldegrasse) [inline samples/modules/lvgl/demos/boards/da1469x_dk_pro.conf:14]: You might consider setting these values in the board's Kconfig.defconfig, since they will be needed by customers using LVGL as well. https://github.com/zephyrproject-rtos/zephyr/blob/7f5b332b58a9bfbbc6ed7b91e1bfc03f5f0ca7f7/boards/arm/mimxrt1060_evk/Kconfig.defconfig#L61

File: west.yml
[the remark this answers is not in the document]
Comment (ioannis-karachalios) [reply west.yml:1]: OK. Correct, the manifest revision points to LCDC fixes (which will otherwise raise compile errors).

File: drivers/mipi_dbi/mipi_dbi_smartbond.c (hunk 7 above)
Comment (ioannis-karachalios) [reply drivers/mipi_dbi/mipi_dbi_smartbond.c:64]: The compiler should read this symbol from memory. Technically speaking, symbols that are also written within ISR should be marked as volatile. Will fix it.

File: modules/lvgl/Kconfig
Comment (ioannis-karachalios) [reply modules/lvgl/Kconfig:1]: I was unsure so I decided to push these changes here. Will open a separate PR.

File: drivers/mipi_dbi/mipi_dbi_smartbond.c (hunk 5 above)
Comment (ioannis-karachalios) [reply drivers/mipi_dbi/mipi_dbi_smartbond.c:230]: Indeed, reading pixel data should not be needed for most use cases. This is why I am safeguarding the read API so it's up to users to exclude it from building or not. Although read operations are not natively supported by our LCDC IP, they can be performed via the standard SPI bus.

File: drivers/mipi_dbi/mipi_dbi_smartbond.c (hunk 6 above)
Comment (ioannis-karachalios) [reply drivers/mipi_dbi/mipi_dbi_smartbond.c:393]: Understood. In contrast to the majority of MIPI DBI controllers, our LCDC IP requires the `height` and `width` descriptor elements which will implicitly determine the number of bytes to be sent.

File: drivers/mipi_dbi/mipi_dbi_smartbond.c (hunk 8 above)
Comment (ioannis-karachalios) [reply drivers/mipi_dbi/mipi_dbi_smartbond.c:588]: From architecture point of view, our SoC integrates a single LCDC IP block. The only reason that multiple instances should be supported is to cover use cases where more than one display device is to be driven. Note sure if this scenario has a meaning in a real application. @danieldegrasse what do you think?

File: drivers/mipi_dbi/mipi_dbi_smartbond.c (hunk 1 above)
Comment (ioannis-karachalios) [reply drivers/mipi_dbi/mipi_dbi_smartbond.c:81]: Given our `pm_state_set` SoC implementation, this should not work here. Thanks for pointing this out, either way.

**Comment by danieldegrasse** [comment]
@ioannis-karachalios can you take a look at the CI/compliance failures?

**Comment by ioannis-karachalios** [comment]
> @ioannis-karachalios can you take a look at the CI/compliance failures?

Sure. I was just waiting for your complete feedback.

**Comment by danieldegrasse** [comment]
> Sure. I was just waiting for your complete feedback.

Understood. I'll continue giving feedback, but I would start addressing the failures. It is difficult to give a final review when CI is failing, since the contents of the PR will necessarily have to change in order for it to merge.

File: samples/drivers/display/boards/da1469x_dk_pro.conf [hunk 11]
```diff
@@ -0,0 +1,10 @@
+#
+# Copyright (c) 2023 Renesas Electronics Corporation
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+#
+CONFIG_HEAP_MEM_POOL_SIZE=4096
+CONFIG_LOG_MODE_MINIMAL=y
+CONFIG_MIPI_DBI_LOG_LEVEL_DBG=y
```
Comment (danieldegrasse) [inline samples/drivers/display/boards/da1469x_dk_pro.conf:9]: Are the log settings needed for customers running the sample? Further, does the heap size need adjustment? It seems like the default (larger) heap size would work ok.

File: samples/drivers/display/boards/da1469x_dk_pro.overlay [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+ #include <zephyr/dt-bindings/display/ili9xxx.h>
+ #include <zephyr/dt-bindings/gpio/gpio.h>
+
+ / {
+    chosen {
+        zephyr,display = &ili9340;
+    };
+ };
+
+ &mipi_dbi {
```
Comment (danieldegrasse) [inline samples/drivers/display/boards/da1469x_dk_pro.overlay:16]: Is the display used here an external module, or integrated with the board? If integrated, it might make sense to define it in the board devicetree instead of as an overlay

Comment (ioannis-karachalios) [reply samples/drivers/display/boards/da1469x_dk_pro.overlay:16]: I was sort of unsure how to handle this as the specific display module is neither integrated with the board nor promoted for the specific display interface. Therefore, I decided to move its configuration in overlay files. However, I will need to safeguard it against other display interfaces/modules as the same LCD controller might also be used to drive [display panels](https://github.com/zephyrproject-rtos/zephyr/pull/67649/commits/31dfb94fc980cbce0db9354f6c92e8f5054ba64c). To be honest, I have not encountered related use cases so, I guess it would be ok to introduce a Kconfig choice in our board's Kconfig.defconfig which should select the display node/module. Let me know if there are other approaches for this to work.

Comment (danieldegrasse) [reply samples/drivers/display/boards/da1469x_dk_pro.overlay:16]: Is this a case where the module could be a [shield](https://docs.zephyrproject.org/latest/hardware/porting/shields.html)? Generally anything using the Arduino interface is a great candidate for this. If the interface is completely nonstandard, keeping the display as an overlay is probably best. NXP has used shields to define displays when there is a shared interface used between multiple display types (we use a 40 pin FPC for several modules) so this could be another option.

Comment (danieldegrasse) [reply samples/drivers/display/boards/da1469x_dk_pro.overlay:16]: To be clear here- the LCD controller pins are exposed on the arduino header, in the same location as the SPI pins are usually exposed? I'm not certain what the best way to handle this is, because the `adafruit_2_8_tft_touch_v2` shield assumes a standard SPI peripheral is exposed via the SPI pins on the arduino header (so it will use the MIPI DBI SPI driver). This might be a case where using a board overlay is the best we can do.

Comment (ioannis-karachalios) [reply samples/drivers/display/boards/da1469x_dk_pro.overlay:16]: Yes, the LCD controller's pins are exposed in the same location as the SPI pins. So, my understanding is that we should stick with my initial approach, correct? In that case, I would like to ask what is the best way to add support for multiple LCD interfaces within the same `<BOARD>.overlay` file. A tried to use Kconfig symbols along with pre-processor directives but it seems that Kconfigs are not interpreted correctly (not sure if it's my fault or it's that such a scheme is not allowed in overlays.)

File: drivers/mipi_dbi/mipi_dbi_smartbond.c (hunk 6 above)
Comment (danieldegrasse) [reply drivers/mipi_dbi/mipi_dbi_smartbond.c:412]: Sure, I follow that- to be clear, what I am suggesting here is that we add a check like the following:

```
if (desc->width * desc->height * (DISPLAY_BITS_PER_PIXEL(pixfmt) / 8) != desc->buf_size) {
    LOG_ERR("Incorrect buffer size for given width and height");
    return -EINVAL;
}
```

This way a user can't pass in a display buffer descriptor that is otherwise invalid in the MIPI DBI API, and have it work with this driver but not with other implementations

Comment (ioannis-karachalios) [reply drivers/mipi_dbi/mipi_dbi_smartbond.c:412]: Clear. I have fixed it at the beginning of the API.

**Comment by ioannis-karachalios** [comment]
@danieldegrasse I am done with all the changes requested. Let me know if more updates are required.

**Comment by ioannis-karachalios** [comment]
#69410 is also required so the LCD controller can handle partial updates appropriately.

File: drivers/mipi_dbi/mipi_dbi_smartbond.c [hunk 13]
```diff
@@ -0,0 +1,620 @@
...
+		return ret;
+	}
+#endif
+
+	return 0;
+}
+
+static struct mipi_dbi_driver_api mipi_dbi_smartbond_driver_api = {
+#if MIPI_DBI_SMARTBOND_IS_RESET_AVAILABLE
+	.reset = mipi_dbi_smnartbond_reset,
```
Comment (danieldegrasse) [inline drivers/mipi_dbi/mipi_dbi_smartbond.c:589]: Nit- typo in this function name

File: boards/renesas/da1469x_dk_pro/da1469x_dk_pro-pinctrl.dtsi [hunk 14]
```diff
@@ -26,20 +26,20 @@
 
 	i2c2_default: i2c2_default {
 		group1 {
-			pinmux = <SMARTBOND_PINMUX(I2C2_SDA, 0, 19)>,
-				<SMARTBOND_PINMUX(I2C2_SCL, 0, 18)>;
+			pinmux = <SMARTBOND_PINMUX(I2C2_SDA, 0, 28)>,
```
Comment (danieldegrasse) [inline boards/renesas/da1469x_dk_pro/da1469x_dk_pro-pinctrl.dtsi:29]: Can you split these pin control fixes into separate commits? It should be clear in the git history why these are needed

File: boards/renesas/da1469x_dk_pro/Kconfig.defconfig [hunk 15]
```diff
@@ -5,8 +5,29 @@
 
 if BOARD_DA1469X_DK_PRO
 
+if LVGL
+
+# LCDC imposes display buffer be word aligned
+config LV_Z_VDB_ALIGN
+	default 4
```
Comment (danieldegrasse) [inline boards/renesas/da1469x_dk_pro/Kconfig.defconfig:12]: This should not be needed- the default value is already 4: https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/lvgl/Kconfig.memory#L66

File: samples/modules/lvgl/demos/boards/da1469x_dk_pro_mipi_dbi.overlay
Comment (danieldegrasse) [inline samples/modules/lvgl/demos/boards/da1469x_dk_pro_mipi_dbi.overlay:1]: Just a suggestion here, since you are reusing a similar overlay to define this display between multiple samples. You might consider putting this overlay in the boards `dts` folder. For example, see how an overlay is included for many NXP boards to switch to a new ethernet driver: https://github.com/zephyrproject-rtos/zephyr/blob/3203bd660c6c8968e3047fb16b5721645f7ca696/boards/nxp/mimxrt1060_evk/dts/nxp%2Cenet-experimental.overlay

File: boards/renesas/da1469x_dk_pro/Kconfig.defconfig (hunk 15 above)
Comment (ioannis-karachalios) [reply boards/renesas/da1469x_dk_pro/Kconfig.defconfig:12]: Correct. The reason I also chose to add it here is because buffer alignment is critical for LCD controller's operation.

**Review by danieldegrasse — COMMENTED** [review]
Sorry- just thought about this. Can you add a testcase entry like the following

https://github.com/zephyrproject-rtos/zephyr/blob/a14ba6a253892903d3e72d10c283adcd2087840c/samples/drivers/display/sample.yaml#L110

So that twister will build this sample with the MIPI DBI driver in the future?

**Review by danieldegrasse — APPROVED** [review]

**Comment by ioannis-karachalios** [comment]
@danieldegrasse many thanks for reviewing this PR.

**Review by andrzej-kaczmarek — APPROVED** [review]
